test/e2e/framework/util.go

```diff
@@ -51,7 +67,78 @@ func ScratchDirs(t TestingTInterface) (string, string, error) {
 		directories = append(directories, dir)
 	}
 	t.Logf("Saving test artifacts and data under %s.", baseTempDir)
-	return directories[0], directories[1], nil
+	t.artifactDir, t.dataDir = directories[0], directories[1]
+	return t.artifactDir, t.dataDir, nil
+}
+
+var localSchemeBuilder = runtime.SchemeBuilder{
+	apiresourcev1alpha1.AddToScheme,
+	tenancyv1alpha1.AddToScheme,
+	clusterv1alpha1.AddToScheme,
+}
+
+func init() {
+	utilruntime.Must(localSchemeBuilder.AddToScheme(scheme.Scheme))
+}
+
+// Artifact runs the data-producing function and dumps the YAML-formatted output
+// to the artifact directory for the test. Normal usage looks like:
+// defer framework.Artifact(t, kcp, client.Get(ctx, name, metav1.GetOptions{}))
+func Artifact(tinterface TestingTInterface, kcpName string, discoveryClient discovery.DiscoveryInterface, producer func() (runtime.Object, error)) {
```

Comment:
Fundamentally I wanted to ask the user for the smallest amount of input to this, which means we end up doing some "extra" things like using the RESTMapper to find a nice name for the artifact and messing with the Scheme so we ensure that we're serializing something sensible to disk that can be `kubectl apply`-ed to another `kcp` later.

Reply:
After a bit more thought, this might make more sense to hang off of the `RunningServer` since that knows how to e.g. get the discovery client for the thing we're talking to, and knows its own name, which are both things that the user should not bother needing to pass in. I might do another refactor pass